README.md
# crosslink-ml-hn  [![Build Status](https://travis-ci.org/liviu-/crosslink-ml-hn.svg?branch=master)](https://travis-ci.org/liviu-/crosslink-ml-hn)

Reddit bot to link from [/r/machinelearning](https://www.reddit.com/r/MachineLearning/) submissions to [HN](https://news.ycombinator.com/news) discussions for same URL submissions to encourage more discussion/engagement between communities.

[/u/hn_crosslinking_bot](https://www.reddit.com/user/hn_crosslinking_bot)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The versioning process follows a scheme somewhat based on [Semantic Versioning](http://semver.org/), and the CHANGELOG.md format is based on [Keep a Changelog](http://keepachangelog.com/).

## [Unreleased]

### Changed
- Comment was changed to link to the issues page rather than the main page of the project to be more in-line with the hyperlink text.
- The footer says "Have a suggestion?" instead of "Report a bug" to encourage people to suggest improvements even if there are no bugs that they wish to report (triggered by [this](https://www.reddit.com/r/MachineLearning/comments/55rdsj/google_research_how_robots_can_acquire_new_skills/d8d9gnf)).

## [1.0.0] - 2016-09-21

### Added
- Added a link to th GitHub page in case people want to report issues -- bugs happen and I'm not always paying attention ([issue #10](https://github.com/liviu-/crosslink-ml-hn/issues/10)).

### Changed
- **BREAKING** Not an immediately breaking change, but only Python3.5 will be tested for by Travis ([issue #17](https://github.com/liviu-/crosslink-ml-hn/issues/17))
- Changed where shell arguments are parsed (at the moment, just `--version`). This makes it slower, but avoids a bug with interactin with `pytest`'s own arguments.

## [0.2.0] - 2016-09-15

### Changed
- Simplified the URL matching scheme. The only exception right now is arXiv as DeepMind updated their blog and adding too many exceptions makes for cumbersome code. If more false positives arise, I'll try to improve then.
- Config file is now stored in YAML format, and the path to the file is by default `$HOME/.crosslinking_config.yaml`, but it may be provided by the user via the `CROSSLINKING_CONFIG` environment variable.

## [0.1.2] - 2016-09-07

### Fixed
- Added anoter exception for PLOS journals because they also encode the article ID in GET params (issue [#16](https://github.com/liviu-/crosslink-ml-hn/issues/16))

## [0.1.1] - 2016-08-29

### Fixed
-- Added `config.py` back to pass CI tests

## [0.1.0] - 2016-08-29

### Fixed
- If the path is simply 'blog', then the URL fragments are considered when differentiating between the links. This accommodates DeepMind URL scheme and hopefully others as well.

### Added
- Added a `--version` CLI argument to the tool
- Added some special rules for arXiv URLs to match slightly different links

## [0.0.2] - 2016-08-29

### Fixed
- Fragment identifiers were not normalized out when comparing 2 URLs (see [issue #8](https://github.com/liviu-/crosslink-ml-hn/issues/8)), causing an issue where `site.com/path#fragment` was not marked as being the same resource as `site.com/path`.

## 0.0.1 - 2016-08-20

Initial release


[1.0.0]: https://github.com/liviu-/crosslink-ml-hn/compare/v0.2.0...v1.0.0
[0.2.0]: https://github.com/liviu-/crosslink-ml-hn/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/liviu-/crosslink-ml-hn/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/liviu-/crosslink-ml-hn/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/liviu-/crosslink-ml-hn/compare/v0.0.2...v0.1.0
[0.0.2]: https://github.com/liviu-/crosslink-ml-hn/compare/v0.0.1...v0.0.2


tests/test_same_url.py
from crosslinking_bot import utils

def test_same_url_even_if_fragment_differs():
    url1 = 'https://example.com/somtehing#nothing'
    url2 = 'https://example.com/somtehing'
    assert utils.same_url(url1, url2) is True

def test_different_urls():
    url1 = 'https://example1.com/somtehing#nothing'
    url2 = 'https://example2.com/somtehing'
    assert utils.same_url(url1, url2) is False

def test_arxiv_same_resource_abs_pdf():
    url1 = 'https://arxiv.org/abs/1608.03282'
    url2 = 'https://arxiv.org/pdf/1608.03282.pdf'
    assert utils.same_url(url1, url2) is True

def test_arxiv_same_resource_ftp_abs():
    url1 = 'https://arxiv.org/ftp/arxiv/papers/1608/1608.03282.pdf'
    url2 = 'https://arxiv.org/abs/1608.03282'
    assert utils.same_url(url1, url2) is True

def test_arxiv_same_resource_abs_pdf_different_versions():
    # It's questionable whether those 2 refer to the same resource
    url1 = 'https://arxiv.org/abs/1608.03282'
    url2 = 'https://arxiv.org/pdf/1608.03282v2.pdf'
    assert utils.same_url(url1, url2) is True

def test_arxiv_same_resource_pdf_different_versions():
    # It's questionable whether those 2 refer to the same resource
    url1 = 'https://arxiv.org/pdf/1608.03282.pdf'
    url2 = 'https://arxiv.org/pdf/1608.03282v2.pdf'
    assert utils.same_url(url1, url2) is True

def test_plos_different_resource_same_base_query():
    url1 = 'http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004485'
    url2 = 'http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004961'
    assert utils.same_url(url1, url2) is False

def test_plos_same_resource():
    url1 = 'http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004485'
    url2 = 'http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004485'
    assert utils.same_url(url1, url2) is True

def test_plos_same_resource_different_encoding():
    url1 = 'http://journals.plos.org/plosone/article?id=10.1371%2Fjournal.pone.0141854'
    url2 = 'http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0141854'
    assert utils.same_url(url1, url2) is True

def test_different_resource_same_base_uses_id_query():
    url1 = 'http://journals.plos.org/article?id=1000'
    url2 = 'http://journals.plos.org/plosone/article?id=1000'
    assert utils.same_url(url1, url2) is False

def test_different_resource_same_base_uses_other_query():
    url1 = 'http://youtube.com/watch?v=something'
    url2 = 'http://youtube.com/watch?v=something-else'
    assert utils.same_url(url1, url2) is False


tests/test_integration.py
from subprocess import Popen, PIPE


TOOL = 'crosslinking_bot'


def run_tool(args=[]):
    p = Popen([TOOL] + args, stdout=PIPE, stderr=PIPE)
    stdout, stderr = (out.decode('utf-8') for out in p.communicate())
    return p, stdout, stderr

def test_simple_call():
    p, stdout, stderr = run_tool()
    assert p.returncode == 0




tests/test_crosslinking_bot.py
from datetime import datetime
from datetime import date, timedelta

import pytest

from crosslinking_bot import crosslinking_bot as cb
from crosslinking_bot import utils


# Test `cb.parse_date()`

def test_return_today():
    today = datetime.today().date()
    assert 'today' == cb.parse_date(today)

def test_return_1_day_ago():
    yesterday = date.today() - timedelta(1)
    assert '1 day ago' == cb.parse_date(yesterday)

def test_return_2_days_ago():
    two_days_ago = date.today() - timedelta(2)
    assert '2 days ago' == cb.parse_date(two_days_ago)


# Test `cv.prepare_comment()`

@pytest.fixture
def hn_hits():
    return [{
        'objectID': 12135399,
        'created_at_i': 1469823139,
        },
        {
        'objectID': 12135398,
        'created_at_i': 1469821139,
        },
    ]

def test_one_hit_contains_right_url(hn_hits):
    hn_hits = [hn_hits[0]]
    hn_url = cb.HN_STORY.format(hn_hits[0]['objectID'])
    assert hn_url in cb.prepare_comment(hn_hits)

def test_two_hits_contain_second_url(hn_hits):
    hn_url = cb.HN_STORY.format(hn_hits[1]['objectID'])
    assert hn_url in cb.prepare_comment(hn_hits)

def test_two_hits_contain_plural_form(hn_hits):
    hn_url = cb.HN_STORY.format(hn_hits[1]['objectID'])
    assert 'discussions' in cb.prepare_comment(hn_hits)

def test_one_hit_github_url_exists(hn_hits):
    hn_hits = [hn_hits[0]]
    hn_url = cb.HN_STORY.format(hn_hits[0]['objectID'])
    assert 'github' in cb.prepare_comment(hn_hits)

def test_two_hits_github_url_exists(hn_hits):
    hn_url = cb.HN_STORY.format(hn_hits[1]['objectID'])
    commmm = (cb.prepare_comment(hn_hits))
    assert 'github' in cb.prepare_comment(hn_hits)

def test_one_hit_github_url_points_to_issues(hn_hits):
    hn_hits = [hn_hits[0]]
    hn_url = cb.HN_STORY.format(hn_hits[0]['objectID'])
    assert 'issues' in cb.prepare_comment(hn_hits)


main.py
#!/usr/bin/env python3
"""Reddit bot that links to HN submissions

The bot checks the top submissions on /r/machinelearning
and compares the links against all HN submissions using the
Algolia API. The bot posts a new top-level reply if there is 
a relatively active match and links to the HN thread.
"""

import re
import time
import collections
from datetime import datetime

import praw
import requests
import urltools

from config import REDDIT_USERNAME, REDDIT_PASS, USER_AGENT

REDDIT_LIMIT = 50

HN_ALGOLIA = 'http://hn.algolia.com/api/v1/search?query={}&restrictSearchableAttributes=url' 
HN_STORY = 'https://news.ycombinator.com/item?id={}'

SLEEP_TIME = 60
# Minimum number of comments of an HN story to be considered
COMM_NUM_THRESHOLD = 3


def get_reddit_submissions():
    """Download top `REDDIT_LIMIT` submissions from the /r/machinelearning subreddit.

    Uses Reddit API to get the hottest submissions. Expects `USER_AGENT`,
    `REDDIT_USERNAME`, and `REDDIT_PASS` in the function scope.

    Returns:
	list: List of relevant praw.objects.Submission.
    """
    r = praw.Reddit(user_agent=USER_AGENT)
    r.login(REDDIT_USERNAME, REDDIT_PASS, disable_warning=True)
    submissions = r.get_subreddit('machinelearning').get_hot(limit=REDDIT_LIMIT)
    return [sub for sub in submissions if 'reddit.com' not in sub.url]


def get_common_submissions(reddit_submissions, min_comments=COMM_NUM_THRESHOLD):
    """Filter common HN and /r/machinelearning submissions.

    The function queries the Algolia API for each URL
    in the submissions provided and fetches the HN IDs for submissions 
    that have more than `min_comments` comments.

    Args:
        reddit_submissions (iter): Iterable containing `praw` submission objects.
        min_comments (int, optional) Minimum number of comments to consider a HN
            submission. The default value is taken from the module constant 
            `COMM_NUM_THRESHOLD`.

    Returns:
        dict: A dict mapping `praw` submission objects to HN hit objects
    """
    common_subs = collections.defaultdict(list)

    for reddit_sub in reddit_submissions:
        for hit in requests.get(HN_ALGOLIA.format(reddit_sub.url)).json().get('hits'):
            try:
                if hit['num_comments'] > COMM_NUM_THRESHOLD and urltools.compare(hit['url'], reddit_sub.url):
                    common_subs[reddit_sub].append(hit)
            # `hit['num_comments'] may return `None`
            except TypeError:
                continue
    return common_subs

def parse_date(date):
    """Parse date to human readable format

    Args:
        date: `datetime.date` object
    Returns:
        str: human readable relative time
    """
    today = datetime.today().date()
    if date == today:
        return 'today'
    else:
        days_ago = (today - date).days
        plural_suffix = 's' if days_ago > 1 else ''
        return '{} day{} ago'.format(days_ago, plural_suffix)


def prepare_comment(hn_hits):
    """Format the comment from the HN hits

    Comments may get more complex when there are multiple 
    hits, so this function tries to format it neatly.

    Args:
        hn_hits: List of dictionaries containing data
            about the HN hits.

    Returns:
        str: Formatted comment.
    """
    header = 'HN discussion: '
    if len(hn_hits) == 1:
        return header + HN_STORY.format(hn_hits[0]['objectID'])
    else:
        # Change the header to use plural form
        header = re.sub(':', 's:', header)
        header += '\n\n'

        hit_strings = []
        for hit in hn_hits:
            hit_date = datetime.fromtimestamp(hit['created_at_i']).date()
            hit_date_human = parse_date(hit_date)
            url = HN_STORY.format(hit['objectID']) 
            hit_strings.append('{} ({})'.format(url, hit_date_human))

    return header + '\n\n'.join(hit_strings)


def post_comments(common_subs):
    """Post comments on Reddit.

    Posts `len(common_subs)` comments every `SLEEP_TIME` seconds. 

    Args:
        common_subs (dict): Maps `praw` submission objects to HN hit objects.
    """
    for reddit_obj, hn_hits in common_subs.items():
	# If I didn't post here before
        if not any(comm for comm in reddit_obj.comments if REDDIT_USERNAME in str(comm.author)):
            comment = prepare_comment(hn_hits)
            reddit_obj.add_comment(comment)
            time.sleep(SLEEP_TIME)

def main():
    reddit_subs = get_reddit_submissions()
    common_subs = get_common_submissions(reddit_subs)
    post_comments(common_subs)

if __name__ == '__main__':
    main()


crosslinking_bot/utils.py
import re
import os

import yaml
import urltools
from urllib import parse

# Environment variable pointing to the config file
config_environ = 'CROSSLINKING_CONFIG'
default_path = os.path.join(os.path.expanduser("~"), '.crosslinking_config.yaml')

def same_url(raw_url1, raw_url2):
    """Check if 2 URLs refer to the same primary resource

    `urltools.compare()` fails if the 2 URLs have different fragments.
    See issue #8 for details. The function treats a special case where
    the path is simply '/blog' to accommodate some blogs that refer to
    their posts via the fragment.

    Args:
        url1 (str): First URL to be compared
        url2 (str): Second URL

    Returns:
        bool: Whether the URLs are the same
    """
    arxiv_exception = 'arxiv.org'
    fragment_identifier = '#'

    url1 = _parse_url(raw_url1)
    url2 = _parse_url(raw_url2)

    # If it's on arxiv, do some acrobatics
    if url1['netloc'] == url2['netloc'] == arxiv_exception:
        regex = '([^/a-z]+\.[^/a-z.]+)'
        return re.findall(regex, url1['path']) == re.findall(regex, url2['path'])
    else:
        return urltools.compare(_normalize_url(raw_url1), _normalize_url(raw_url2))

def _parse_url(url):
    """Parse URL into a dictionary"""
    url_dict = {}

    parsed_result = parse.urlparse(url)
    for key, value in zip(parsed_result._fields, parsed_result):
        url_dict[key] = value if key != 'query' else parse.parse_qs(value)

    return url_dict

def _normalize_url(url):
    """Remove fragment from an URL"""
    return url.split('#', 1)[0]

def get_config():
    """Read config file and return Python dictionary"""
    config_file = os.environ.get(config_environ) or default_path
    with open(config_file) as f:
        config = yaml.load(f)
    return config 



crosslinking_bot/sources.py
"""Sources to check for common links

This module constains different discussion sources that the bot uses
to find potential relevant links to crosslink.
"""
import collections
from abc import ABCMeta, abstractmethod

import requests

from . import utils

MIN_ACTIVITY = 3


class Source(object):
    """Abstract clas for discussion sources to subclass
    
    There is some common functionality between different
    sources, so this serves as a starting point for them.
    """
    __metaclass__ = ABCMeta

    def __init__(self, reddit_submissions, min_activity=MIN_ACTIVITY):
        """Constructor to store shared variables for different sources

        Args:
            self.reddit_submissions (iter): Iterable containing `praw` submission objects.
            self.min_comments (int, optional) Minimum number of comments to consider a 
                submission.
        """
        self.reddit_subs = reddit_submissions
        self.min_activity = min_activity

    @abstractmethod
    def get_common_submissions(self):
        pass


class HN(Source):
    """Hacker News source

    Args:
        _hn_algolia (str): Links to the API used for retrieving
            HN common submissions
    """
    _hn_algolia = ('http://hn.algolia.com/api/v1/search?'
                   'query={}&restrictSearchableAttributes=url')

    def get_common_submissions(self):
        """Filter common HN and /r/machinelearning submissions.

        The method queries the Algolia API for each URL
        in the submissions provided and fetches the HN IDs for submissions 
        that have more than `min_comments` comments.


        Returns:
            dict: A dict mapping `praw` submission objects to HN hit objects
        """
        common_subs = collections.defaultdict(list)

        for reddit_sub in self.reddit_subs:
            for hit in requests.get(self._hn_algolia.format(reddit_sub.url)).json().get('hits', []):
                try:
                    if (hit['num_comments'] > self.min_activity and 
                        utils.same_url(hit['url'], reddit_sub.url)):
                        common_subs[reddit_sub].append(hit)
                # `hit['num_comments'] may return `None`
                except TypeError:
                    continue
        return common_subs

class RedditStats(Source):
    """/r/statistics source """

    def get_common_submissions(self):
        pass


crosslinking_bot/crosslinking_bot.py
#!/usr/bin/env python3
"""Reddit bot that links to HN submissions

The bot checks the top submissions on /r/machinelearning
and compares the links against all HN submissions using the
Algolia API. The bot posts a new top-level reply if there is 
a relatively active match and links to the HN thread.
"""

import re
import time
import collections
import logging
from datetime import datetime

import praw
import requests

from .sources import HN
from .utils import get_config
from . import parse_args

CONFIG = get_config()
DEBUG = False

REDDIT_LIMIT = 50

HN_ALGOLIA = 'http://hn.algolia.com/api/v1/search?query={}&restrictSearchableAttributes=url' 
HN_STORY = 'https://news.ycombinator.com/item?id={}'

SLEEP_TIME = 60
# Minimum number of comments of an HN story to be considered
COMM_NUM_THRESHOLD = 3

logging.basicConfig(filename='.log',level=logging.DEBUG)

def get_reddit_submissions():
    """Download top `REDDIT_LIMIT` submissions from the /r/machinelearning subreddit.

    Uses Reddit API to get the hottest submissions. Expects `CONFIG`
    in the function scope.

    Returns:
	list: List of relevant praw.objects.Submission.
    """
    r = praw.Reddit(user_agent=CONFIG['agent'])
    r.login(CONFIG['user'], CONFIG['pass'], disable_warning=True)
    submissions = r.get_subreddit('machinelearning').get_hot(limit=REDDIT_LIMIT)
    return [sub for sub in submissions if 'reddit.com' not in sub.url]


def get_common_submissions(reddit_submissions, min_activity=COMM_NUM_THRESHOLD):
    """Get common submissions from various sources

    The function finds submissions to crosslink that are relevant to
    the submissoins provided. At the moment, only Hacker News is 
    considered.

    Args:
        reddit_submissions (iter): Iterable containing `praw` submission objects.
        min_activity (int, optional) Minimum number of comments to consider a
            submission. Default value is taken from the module-level constant
            `COMM_NUM_THRESHOLD`.
    Returns:
        dict: A dict mapping `praw` submission objects to HN hit objects
    """
    hn = HN(reddit_submissions, min_activity)
    return hn.get_common_submissions()

def parse_date(date):
    """Parse date to human readable format

    Args:
        date: `datetime.date` object
    Returns:
        str: human readable relative time
    """
    today = datetime.today().date()
    if date == today:
        return 'today'
    else:
        days_ago = (today - date).days
        plural_suffix = 's' if days_ago > 1 else ''
        return '{} day{} ago'.format(days_ago, plural_suffix)


def prepare_comment(hn_hits):
    """Format the comment from the HN hits

    Comments may get more complex when there are multiple 
    hits, so this function tries to format it neatly.

    Args:
        hn_hits: List of dictionaries containing data
            about the HN hits.

    Returns:
        str: Formatted comment.
    """
    header = 'HN discussion: '
    footer = ("\n\n - - - \n\n"  
              "[Have a suggestion?](https://github.com/liviu-/crosslink-ml-hn/issues)")

    if len(hn_hits) == 1:
        return header + HN_STORY.format(hn_hits[0]['objectID']) + footer
    else:
        # Change the header to use plural form
        header = re.sub(':', 's:', header)
        header += '\n\n'

        hit_strings = []
        for hit in hn_hits:
            hit_date = datetime.fromtimestamp(hit['created_at_i']).date()
            hit_date_human = parse_date(hit_date)
            url = HN_STORY.format(hit['objectID']) 
            hit_strings.append('{} ({})'.format(url, hit_date_human))

    return header + '\n\n'.join(hit_strings) + footer


def post_comments(common_subs):
    """Post comments on Reddit.

    Posts `len(common_subs)` comments every `SLEEP_TIME` seconds. 

    Args:
        common_subs (dict): Maps `praw` submission objects to HN hit objects.
    """
    for reddit_obj, hn_hits in common_subs.items():
	# If I didn't post here before
        if not any(comm for comm in reddit_obj.comments if CONFIG['user'] in str(comm.author)):
            comment = prepare_comment(hn_hits)
            reddit_obj.add_comment(comment)
            logging.info(comment)
            time.sleep(SLEEP_TIME)

def run_bot():
    parse_args()
    reddit_subs = get_reddit_submissions()
    common_subs = get_common_submissions(reddit_subs)
    post_comments(common_subs)
